symengine/ntheory.cpp

```diff
@@ -1325,6 +1342,66 @@ bool _nthroot_mod_prime_power(std::vector<RCP<const Integer>> &roots, const inte
     return true;
 }
 
+// Returns whether Solution for x**n == a mod p**k exists or not
+bool is_nthroot_mod_prime_power(const integer_class &a, const integer_class &n, const integer_class &p, const unsigned k)
+{
+            integer_class pk;
+    if (a % p != 0) {
+        if (p == 2) {
+            integer_class t;
+            unsigned c = mp_scan1(n);
+
+            // Handle special cases of k = 1 and k = 2.
+            if (k == 1) {
+                return true;
+            }
+            if (k == 2) {
+                if (c > 0 and a % 4 == 3) {
+                    return false;
+                }
+                return true;
+            }
+            if (c >= k - 2) {
+                c = k - 2;  // Since x**(2**c) == x**(2**(k - 2)) mod 2**k, let c = k - 2.
+            }
+            if (c == 0) {
+                // x**r == a mod 2**k and x**2**(k - 2) == 1 mod 2**k, implies x**(r * s) == x == a**s mod 2**k.
+                return true;
+            }
+
+            // First, solve for y**2**c == a mod 2**k where y == x**r
+            t = integer_class(1) << (c + 2);
+            mp_fdiv_r(t, a, t);
+            // Check for a == y**2**c == 1 mod 2**(c + 2).
+            if (t != 1)
+                return false;
+            return true;
+        } else {
+            return is_nthroot_mod1(a, n, p, k);
+        }
+    } else {
+        integer_class _a;
```

Comment:
See this from wikipedia
> In general, to determine if a quadratic congruence with composite modulus is solvable use the following theorem:[37]

> Let n > 1, and gcd(a,n) =1. Then x2 ≡ a (mod n) is solvable if and only if:

> The Legendre symbol \left(\frac{a}{p}\right)=1 for all odd prime divisors p of n.
a ≡ 1 (mod 4) if 4|n, but 8 does not divide n; a ≡ 1 (mod 8) if 8|n.